### PR TITLE
(#17) Autosuggest implementation

### LINKED ISSUE #17: Story: Autosuggest
##  As a user, I would like to have suggestions appear when I type in the search box 

> **ESTIMATE** 8

### Acceptance Criteria

```gherkin

Background:
  Given "language" is set to "en"
  And "dictionaryName" is set to "Cancer.gov"

Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected "Starts With"
Given user is on landing page for the selected Dictionary
  Then “Starts with” is selected by default 
    And search bar is displayed below introductory text 
    And search bar contains a placeholder text “Enter keywords or phrases” 
  When user types "cancer" in the search bar
  Then placeholder text disappears  
    And Autosuggest appears after user types in 3 or more characters 
  Then highlighting of the text entered appears in the autosuggest field (only highlight if it starts with using the first word) 
```

```gherkin
Background
  Given "language" is set to "es"
  And "dictionaryName" is set to "Cancer.gov"

Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected "Empieza con"
Given user is on landing page for the selected Dictionary
  Then "Empieza con" is selected by default 
    And search bar is displayed below introductory text 
    And search bar contains a placeholder text "Escriba frase o palabra clave"
 When user types "metastásico" in the search bar
    Then Autosuggest appears after user types in 3 or more characters 
  And highlighting of the text entered appears in the autosuggest field (only highlight if it starts with using the first word) 
  When user selects autosuggested term
  Then term is populated into the search bar
  When user submits their search clicking the "Buscar" button
  Then search is executed
  And URL contains selected term
```

```gherkin
Background:
  Given "language" is set to "en"
  And "dictionaryName" is set to "Cancer.gov"

Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected the Contains option
  Given user is on landing page for the selected Dictionary
  When user selects “Contains” option
  Then search bar is displayed below introductory text 
    And search bar contains a placeholder text "Enter keywords or phrases"
  When user types "Breast Cancer" in the search bar
  Then placeholder text disappears  
    And Autosuggest appears after user types in 3 or more characters 
  Then highlighting of the text entered appears in the autosuggest field (highlight any place of word, rather than beginning of word/phrase) 
  When user selects autosuggested term
  Then term is populated into the search bar
   When user submits their search clicking the "Search" button
  Then search is executed 
And URL contains selected term
```

```gherkin
Background
  Given "language" is set to "es"
  And "dictionaryName" is set to "Cancer.gov"

Scenario: User is able to see autosuggested items when typing 3 or more characters into the search box and has selected the Contains option in Spanish
   Given user is on landing page for the selected Dictionary
  When user selects “Contiene” option
  Then search bar is displayed below introductory text 
    And search bar contains a placeholder text “Escriba frase o palabra clave" 
 When user types "metasta" in the search bar
  Then placeholder text disappears  
    And Autosuggest appears after user types in 3 or more characters 
  Then highlighting of the text entered appears in the autosuggest field (highlight any place of word, rather than beginning of word/phrase) 
  When user selects autosuggested term
  Then term is populated into the search bar
  When user submits their search clicking the "Buscar" button
  Then search is executed 
  And URL contains selected term
```

```gherkin
 Given "language" is set to "en"
  And "dictionaryName" is set to "Cancer.gov"

Scenario: User can execute search by clicking Search button without selecting an option from autocomplete
 Given user is on landing page for the selected Dictionary
  When user types "cancer"  in the search bar 
  Then placeholder text disappears  
    And Autosuggest appears after user types in 3 or more characters 
  Then highlighting of the text entered appears in the autosuggest field (only highlight if it starts 
  with using the first word) 
  When user submits their search clicking the "Search" button
 Then search is executed
And URL contains searched term
```

```gherkin
 Given "language" is set to "en"
  And "dictionaryName" is set to "Cancer.gov"

Scenario: User can execute search by clicking Search button without selecting an option from autocomplete
 Given user is on landing page for the selected Dictionary
    And user selects "Contains" option
  When user types "cancer"  in the search bar 
  Then placeholder text disappears  
    And Autosuggest appears after user types in 3 or more characters 
  Then highlighting of the text entered appears in the autosuggest field 
When user submits their search clicking the "Search" button
  Then Search is executed 
And URL contains searched term
```

```gherkin
Scenario: As a user, I would like to see an option to search when I am viewing the definition for a term
  Given the user is viewing the definition with the pretty url "metastatic"
  Then the page title is "metastatic"
    And the pronunciation text "(meh-tuh-STA-tik)" appears on the page
    And the definition text "Having to do with metastasis, which is the spread of cancer from the primary site (place where it started) to other places in the body." appears on the page
And there should be a CGOV Image at position 1 with the following
         | src     | https://nci-media-dev.cancer.gov/pdq/media/images/764135-571.jpg                                                                                                                                                                                                                                 |
         | caption | Metastasis. In metastasis, cancer cells break away from where they first formed (primary cancer), travel through the blood or lymph system, and form new tumors (metastatic tumors) in other parts of the body. The metastatic tumor is the same type of cancer as the primary tumor.            |
         | href    | https://nci-media-dev.cancer.gov/pdq/media/images/764135.jpg                                                                                                                                                                                                                                     |
         | alt     | Metastasis; drawing shows primary cancer that has spread from the colon to other parts of the body (the liver and the lung). An inset shows cancer cells spreading from the primary cancer, through the blood and lymph system, to another part of the body where a metastatic tumor has formed. |
And a video is displayed with the following
         | video_id | fQwar_-QdiQ                                                                                                                                                                                                                                                                    |
         | title    | Metastasis: How Cancer Spreads                                                                                                                                                                                                                                                 |
         | caption  | Many cancer deaths are caused when cancer moves from the original tumor and spreads to other tissues and organs. This is called metastatic cancer. This animation shows how cancer cells travel from the place in the body where they first formed to other parts of the body. |
  Then heading "Search NCI's Dictionary of Cancer Terms" appears
  Then search options for "Starts with" and "Contains" appears
  Then "keywords" search box appears
  Then search button appears beside search box with "Search"
  Then "Browse:" appears with A-Z List of Links beside it
  When user types "Breast Cancer" in the search bar
  Then placeholder text disappears  
    And Autosuggest appears after user types in 3 or more characters 
  When user selects autosuggested term
  Then term is populated into the search bar
   When user submits their search clicking the "Buscar" button
  Then search is executed 
And URL contains selected term
```

```gherkin
Background:
  Given "language" is set to "es"

Scenario: As a user, I would like to see an option to search when I am viewing the definition for a term in Spanish
  Given the user is viewing the definition with the pretty url "cancer-terminal"
 Then the page title is "cáncer terminal"
     And the pronunciation text does not appear on the page
     And the definition text "Cáncer que no se puede curar y que es mortal. También se llama cáncer en estadio terminal." appears on the page
 Then there is a heading on the page with "Más información"
        And there is an resource item with the following link and text
            | https://www.cancer.gov/espanol/cancer/cancer-avanzado/opciones-de-cuidado/hoja-informativa-cuidado | Los cuidados al final de la vida de personas con cáncer                                    |
            | https://www.cancer.gov/espanol/cancer/cancer-avanzado/planes/etapa-final-pdq                       | Planificar la transición para la atención del cáncer avanzado en la etapa final de la vida |
  Then heading "Buscar en el Diccionario de cáncer" appears
   And "keywords" search box appears
    And search button appears beside search box with "Buscar"
    And "Ojear:" appears with A-Z List of Links beside it
When user types "meta" in the search bar
   When user submits their search clicking the "Buscar" button
  Then search is executed 
And URL contains searched term
```

### Resources:
* Mockups: https://publicis.invisionapp.com/share/HZVRID6RSJ7#/screens/402838500
* Web Page: https://www.cancer.gov/publications/dictionaries/cancer-terms/def/metastasis

### Notes
* Since we will have multiple titles based on whichever dictionary is being viewed, and Spanish is complicated we will need to make it so that we pass in the text above the search bar to the initialization of the app. 
---

## Solution

### Prerequisites
- [x] Basic Term Page view has been built (#15)
- [x] Basic Search Box has been built (#76)
- [x] A-Z list has been built (#76)
- [x] Starts With and Contains buttons have been built (first round - no autosuggest) (#76)

### Tasks
- [ ] Add `Autosuggest` component #128 
- [ ] Invoke autosuggest call from `Search` component
- [ ] Update `Search` component test
- [ ] Integrate autosuggest into Search component
- [x] Write automated tests

### Technical Notes
- More notes, or technical nuances/direction


### Resources:
* Feature: https://tracker.nci.nih.gov/browse/CGOVWCMS-30

### Notes
- This should work for both English and Spanish dictionaries

---

## Solution

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Basic Search Function (#14)

### Tasks
- [ ] Short Task Description - Issue #9999
- [ ] Short Task Description - Issue #9999

### Technical Notes
- More notes, or technical nuances/direction

### LINKED ISSUE #128: Autosuggest component
*Add Autosuggest component*

**Parent Issue:** #17 

### Sub-Tasks
- [x] Create Component
- [ ] Implement Unit Tests for Component

### PR DESCRIPTION
**Pull request by seyilonge-nci**
Closes #17
Closes #128 

* Added Autosuggest component and dependencies
* Integrate Autosuggest with Search
* Removed console warnings when exception is raised as expected for customFetch test
* Fixed issue with url parameters not being picked up by search box
* Fixed URL decode issue on AZ list for #
* Fixed issue for search match type toggle for autosuggest
* Set signal in request options to null to stop throttle abort requests, set selected match type option when url parameter changes to updated value
* Added search text substring highlighting to Autocomplete options
* Added bold highligting for autosuggest options
* Updates to fix failing unit and integration tests
* Upping code coverage...
* Added integration tests
* Updated unit tests for Autocomplete component (WIP...)
* Updates to Autocomplete unit tests, fixed act warning in console for Search unit tests with async events
* Resolved useEffect dependency warning
* Updated Autocomplete test suite
* Added Spanish translations for Autocomplete input messages
* Fixed url encoding for searchText parameter in input
* Reverted cleanup on files not related to PR
* Additional code clean up
* Added additional tests for Autocomplete
* Updated Search component test
* Added __snapshots__ dir to .gitignore
* Updated name on RemovableTag test file
* Updated App route tests to only test location without rendering
* Dropped statement coverage for tests (THIS SHOULD BE A TEMPORARY MEASURE)
* Added Autocomplete snapshot directory
* Added aria-selected to option for Autocomplete

### CODE REVIEW

**Comment by seyilonge-nci** [comment]
@VictoriaSunNIH @lburack @blilianyu 
Please review https://react-app-dev.cancer.gov/glossary-app/pr-131
Do try to tinker with the url parameters to hit the app and see the required results as well

**Comment by blilianyu** [comment]
@seyilonge-nci I'm not seeing translated text for auto suggest. I will ask the client for some text.

**Comment by seyilonge-nci** [comment]
@blilianyu True, translated text not available for items listed below:
- [ ] No results found
- [ ] Please enter 3 or more characters

Please feel free to add any others you see. Thanks.

**Comment by blilianyu** [comment]
@seyilonge-nci Translations are below!

- [ ] Ingrese 3 o más caracteres
- [ ] No hay resultados para su búsqueda

**Comment by seyilonge-nci** [comment]
> @seyilonge-nci Translations are below!
> 
> * [x]  No hay resultados para su búsqueda
> * [x]  Ingrese 3 o más caracteres

Thanks @blilianyu! Updated!

**Comment by blilianyu** [comment]
@seyilonge-nci Looks great!!

File: src/__tests__/App.test.js [hunk 1]
```diff
@@ -47,10 +47,13 @@ describe("App component", () => {
   ]);
 
   beforeAll( () => {
+    jest.spyOn(console, 'error');
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/__tests__/App.test.js:50]: This was a stop gap measure to get this in. Not going to be in the PR for review as it's not ready for review yet :-)

File: cypress/integration/common/MetaTags.js [hunk 2]
```diff
@@ -2,13 +2,22 @@
 import { Then } from 'cypress-cucumber-preprocessor/steps';
 
 Then('the page contains meta tags with the following names', dataTable => {
-  for (const { name, content } of dataTable.hashes()) {
+  cy.document().then(doc => {
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply cypress/integration/common/MetaTags.js:5]: This was done to fix an intermittently failing test for an async action.

File: src/components/atomic/Autocomplete/__tests__/Autocomplete.tests.js [hunk 3]
```diff
@@ -0,0 +1,376 @@
+import { cleanup, fireEvent, render, screen } from "@testing-library/react";
+import React from "react";
+
+import Autocomplete from "../Autocomplete";
+
+let inputValue;
+const label = "Mock Label";
+const options = [
+    { key: "movie", name: "Django Unchained" },
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/atomic/Autocomplete/__tests__/Autocomplete.tests.js:9]: :-D Tried to make it fun as writing tests for this was not fun at all. Would update with cancer terms.

Shoot! I did forget some good ones especially Jackie Brown :-)

File: src/views/Definition/Definition.jsx [hunk 4]
```diff
@@ -52,7 +52,7 @@ const Definition = () => {
 
   const renderMetaDefinition = () => {
     const regex = new RegExp(/[^.!?]+[.!?]+/g);
-    var definitionSplit = payload.definition.text.match(regex);
+    let definitionSplit = payload.definition.text.match(regex);
```
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/views/Definition/Definition.jsx:55]: No it does not. Noticed I was up to my usual and have reverted this already. Would be in the next push. The variable gets assigned a value again on `ln 58` hence the need for a `let`.

File: src/components/atomic/RemovableTag/__tests__/Removable.tests.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { fireEvent, render } from "@testing-library/react";
```
Comment (arcepaul) [inline src/components/atomic/RemovableTag/__tests__/Removable.tests.js:1]: why is this filename not `RemovableTag.tests.js`?

Comment (seyilonge-nci) [reply src/components/atomic/RemovableTag/__tests__/Removable.tests.js:1]: Thanks for the catch, that was an omission. Will update that.

File: src/components/molecules/search/Search.jsx [hunk 6]
```diff
@@ -40,31 +58,91 @@ const Search = () => {
...
+                    placeholder: i18n.searchPlaceholderText[language]
+                }}
+                items={ autoSuggest.payload || [] }
+                getItemValue={ item => item.termName }
+                shouldItemRender={ matchItemToTerm }
+                onChange={ event => onChangeHandler( event ) }
+                onSelect={ ( value, item ) => onSelectHandler( value ) }
+                renderMenu={ children => (
+                    <div
+                        className="ncids-autocomplete__menu --terms"
```
Comment (arcepaul) [inline src/components/molecules/search/Search.jsx:110]: 1. Why `--terms` here?
2. Could you also add a `role=listbox` here?

File: src/components/molecules/search/Search.jsx [hunk 7]
```diff
@@ -40,31 +58,91 @@ const Search = () => {
...
+                            )
+                        }
+                    </div>
+                ) }
+                renderItem={ ( item, isHighlighted ) => (
+                    <div
+                        className={`ncids-autocomplete__menu-item ${
+                            isHighlighted ? 'highlighted' : ''
+                        }`}
+                        key={item.termId}
```
Comment (arcepaul) [inline src/components/molecules/search/Search.jsx:136]: Could you add the attribute `role="option"` here?

File: src/components/molecules/search/Search.jsx (hunk 6 above)
Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx:110]: :-) I added it to the test. Will do the same here. Thanks.

File: src/components/molecules/search/Search.jsx (hunk 7 above)
Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx:136]: Sure, will do.

File: src/components/molecules/search/Search.jsx (hunk 6 above)
Comment (arcepaul) [reply src/components/molecules/search/Search.jsx:110]: wait I don't see `role=listbox` anywhere?

File: src/components/molecules/search/Search.jsx (hunk 7 above)
Comment (arcepaul) [reply src/components/molecules/search/Search.jsx:136]: Not as a child of `listbox`.  Children of a listbox should be `option`

Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx:136]: Will update.

File: src/components/molecules/search/Search.jsx (hunk 6 above)
Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx:110]: Sorry, that was a `combobox` and `listbox` will update.

File: cypress/integration/common/MetaTags.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply cypress/integration/common/MetaTags.js:5]: Fixed.

File: src/__tests__/App.test.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/__tests__/App.test.js]: Fixed.

File: src/components/atomic/Autocomplete/__tests__/Autocomplete.tests.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/components/atomic/Autocomplete/__tests__/Autocomplete.tests.js]: Updated.

File: src/views/Definition/Definition.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (seyilonge-nci) [reply src/views/Definition/Definition.jsx]: Reverted.

File: src/components/atomic/RemovableTag/__tests__/Removable.tests.js (hunk 5 above)
Comment (seyilonge-nci) [reply src/components/atomic/RemovableTag/__tests__/Removable.tests.js]: Done.

File: src/components/molecules/search/Search.jsx (hunk 6 above)
Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx:110]: Resolved.

**Comment by seyilonge-nci** [comment]
@bryanpizzillo I've tried to capture most scenarios I can think of for `Autocomplete` component and the coverage still isn't optimal. Test coverage screenshot below. 

Screen Shot 2020-05-08 at 8 12 32 AM

I had to drop the statement threshold in `package.json` for the tests in order for the build to pass from `10` to `50`. When you do have time, please take a look at it. I'm definitely either not doing this right or missed some glaring or subtle use cases.

File: src/components/molecules/search/Search.jsx (hunk 7 above)
Comment (seyilonge-nci) [reply src/components/molecules/search/Search.jsx]: Updated.

**Comment by seyilonge-nci** [comment]
> I gave up on looking for all the extra files that did not need to be changed and commenting. @seyilonge-nci I would like you to look through the 34 files in this commit and see what has ONLY to do with Autosuggest.
> 
> There was one comment regarding the test data and maybe not using Quentin Tarantino movies...

Gone through all the files

**Comment by seyilonge-nci** [comment]
> @bryanpizzillo @arcepaul I've tried to capture most scenarios I can think of for `Autocomplete` component and the coverage still isn't optimal. Test coverage screenshot below.
> 
> Screen Shot 2020-05-08 at 8 12 32 AM
> 
> I had to drop the statement threshold in `package.json` for the tests in order for the build to pass from `10` to `50`. When you do have time, please take a look at it. I'm definitely either not doing this right or missed some glaring or subtle use cases.

`Autocomplete` unit tests have been updated, thanks @sarinapadilla and @kate-mashkina. Coverage has improved considerably but still not optimal and below the previous statement threshold value of 10. I've updated the current threshold value to 20. Screenshot for test coverage below.
Screen Shot 2020-05-13 at 9 38 38 AM

**Comment by bryanpizzillo** [comment]
Autocomplete bolding is broken:

**Bad**
image

**Good**
image

**Comment by bryanpizzillo** [comment]
@seyilonge-nci - Have you rebased of develop yet? it says you still have a conflicting package.json

**Comment by seyilonge-nci** [comment]
> Autocomplete bolding is broken:
> 
> **Bad**
> `gene` is being highlighted in the middle of the word
> image
> 
> **Good**
> image

Saw this and felt I understood where this was coming from for a fix, but wasn't able to reproduce to apply a fix. Can you please state exact scenario? I tried for both "Starts with" and "Contains".

**Comment by seyilonge-nci** [comment]
> @seyilonge-nci - Have you rebased of develop yet? it says you still have a conflicting package.json

I have locally but haven't updated the PR yet 'cos of the scheduled demo. Didn't want to upset the cart in anyway even though it shouldn't affect it.

**Comment by seyilonge-nci** [comment]
@bryanpizzillo Rebased, PR updated! `package.json` conflict resolved.

**Comment by bryanpizzillo** [comment]
> > Autocomplete bolding is broken:
> > **Bad**
> > `gene` is being highlighted in the middle of the word
> > image
> > **Good**
> > image
> 
> Saw this and felt I understood where this was coming from for a fix, but wasn't able to reproduce to apply a fix. Can you please state exact scenario? I tried for both "Starts with" and "Contains". Not getting same data returned.

Use the Genetics Glossary

**Comment by seyilonge-nci** [comment]
> > > Autocomplete bolding is broken:
> > > **Bad**
> > > `gene` is being highlighted in the middle of the word
> > > image
> > > **Good**
> > > image
> > 
> > 
> > Saw this and felt I understood where this was coming from for a fix, but wasn't able to reproduce to apply a fix. Can you please state exact scenario? I tried for both "Starts with" and "Contains". Not getting same data returned.
> 
> Use the Genetics Glossary
> 👍

Fixed!
Screen Shot 2020-05-14 at 10 58 49 AM

**Comment by seyilonge-nci** [comment]
@bryanpizzillo Options items limit of `10` for autosuggest applied according to updated requirements.

**Comment by bryanpizzillo** [comment]
Ok, I am going to pull it in. I found some new issues, but we need to put this to bed. Thanks @seyilonge-nci  for the hard work on this one. 

Oh, and for the record, I also do not like the snapshot tests. C'est la vie.

**Pull request merged**